### PR TITLE
Addresses may not be persistent across calls to pas_enumerator_reader

### PR DESCRIPTION
**Pull request by ddegazio**
#### 8d6284de0e3fef99a91c909e2bb9ffe1c4064f6f
<pre>
Addresses may not be persistent across calls to pas_enumerator_reader
<a href="https://bugs.webkit.org/show_bug.cgi?id=272404">https://bugs.webkit.org/show_bug.cgi?id=272404</a>
<a href="https://rdar.apple.com/125072587">rdar://125072587</a>

Reviewed by Yusuke Suzuki.

Stores the results of calling the reader function in pas_enumerator_create
locally as opposed to referencing the remote memory. This reduction in
indirection means changes to the enumerating process&apos; mapping of the remote
memory will no longer cause values the enumerator is actively using (such
as the pas_root) to become unmapped.

* Source/bmalloc/libpas/src/libpas/pas_enumerator.c:
(pas_enumerator_create):

Canonical link: <a href="https://commits.webkit.org/277271@main">https://commits.webkit.org/277271@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/ca9ff4ffd50bce926ff75b84a5209d88130e2455

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47184 "12 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/26363 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49829 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49867 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43233 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/31695 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23823 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/38434 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47765 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/23853 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/40673 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/19745 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/21263 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/41829 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5228 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/40452 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/43564 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/42237 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/51743 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/46673 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/22211 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/18586 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/45728 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/23486 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/44736 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/10404 "Build was cancelled. Recent messages:OS: Monterey (12.6.8), Xcode: 13.4.1; Checked out pull request; Reviewed by Yusuke Suzuki; Compiled WebKit (warnings); layout-tests (exception)") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/24273 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/54173 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6635 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/23203 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/11118 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") |

### CODE REVIEW

**Review by Constellation — APPROVED** [review]
r=me

File: Source/bmalloc/libpas/src/libpas/pas_enumerator.c [hunk 1]
```diff
@@ -95,39 +95,52 @@ pas_enumerator* pas_enumerator_create(pas_root* remote_root_address,
...
-    if (!result->root)
+    /*
+     * Some non-obvious behavior: even if we expect the remote address of a value (like the
+     * remote pas_root) to be persistent, our local mapping of the remote memory is allowed to
+     * change, and can in fact change any time we call the reader. So, it's unsafe to store
+     * a pointer to the remote memory across a call to reader, and we need to instead allocate
+     * local storage for these values and copy the remote values in.
+     */
+
+    pas_root* remote_root = reader(result, remote_root_address, sizeof(pas_root), reader_arg);
```
Comment (MenloDorian) [inline Source/bmalloc/libpas/src/libpas/pas_enumerator.c]: Is this (declaring a local var in the middle of the function) allowed by C?

**Comment by ddegazio** [comment]
Style failure is expected, since we're touching libPAS. I've manually gone through the errors and all of them are just due to respecting the existing libPAS style conventions.

File: Source/bmalloc/libpas/src/libpas/pas_enumerator.c (hunk 1 above)
Comment (ddegazio) [reply Source/bmalloc/libpas/src/libpas/pas_enumerator.c]: Yes as of C99 or later, which I'm pretty sure we support...but _most_ of libPAS respects this, so I'll refactor it real quick.

Comment (ddegazio) [reply Source/bmalloc/libpas/src/libpas/pas_enumerator.c]: Resolved in latest push.

**Comment by ddegazio** [comment]
Latest push just executes the refactor suggested by Mark above - it's not material, and I've confirmed it still builds locally, so it should be safe to merge without a full EWS run.

**Pull request merged**